up-l1/mqtt.adoc

```diff
@@ -35,8 +35,85 @@ MQTT is an OASIS standard messaging protocol for the Internet of Things (IoT). I
 
 For more information, please refer to https://mqtt.org/
 
-
 == Specifications
 
+=== UAttribute Mapping
+
+MQTT 5 supports custom header fields, and we leverage this to map certain UAttribute values into the MQTT header.
+
+[cols="1,1"]
+|===
+| MQTT Header Property Code | Value (key:value)
+| UserProperty
+| ("upriority","{UAttribute.priority}")
+| CorrelationData
+| ("reqId", "{UAttribute.reqId}")
+| UserProperty
+| ("sink","{UAttribute.sink}") # Should this be serialized to MicroUri?
+|===
+
+The upriority field is needed as mqtt's QoS is very limited, so this will allow listeners to prioritize messages.
+
+The Correlation Data Property Code contains the request correlation id.
+
+The sink field is used to correctly map a message to the correct uE as the topic name may not contain all of the information.
+
+=== URI Mapping
+
+Depending on whether the UURI is local or remote, the MQTT Topic that the message will be published on will be different:
+
+* The remote UUri (with UAuthority): `upr/UUri.UAthority.name`
+* The local UUri (without UAuthority): `upl/UUri.UEntity/UUri.UResource`
+
+Examples:
+
+* The remote UUri (with UAuthority):
+
+[source]
+----
+UUri {
+  authority: UAuthority {
+    name: "Vehicle_1",
+    number: Id({01, 02, 03, 10, 11, 12})
+  },
+  entity: UEntity {
+    name: "body.access",
+    version_major: 1,
+    id: 1234,
+  }
+  resource: UResource {
+    name: "door",
+    instance: "front_left",
+    message: "Door",
+    id: 5678,
+  }
+}
+----
+
+Since this is a remote UUri, the MQTT Topic will simply be `upr/Vehicle_1`.
+
+* The local UUri (without UAuthority):
+
+[source]
+----
+UUri {
+  authority: empty,
+  entity: UEntity {
+    name: "body.access",
+    version_major: 1,
+    id: 1234,
+  }
+  resource: UResource {
+    name: "door",
+    instance: "front_left",
+    message: "Door",
+    id: 5678,
+  }
+}
+----
+
+Since this is a local UUri, the MQTT Topic will be constructed from the UEnity name and the UResource name `upl/body.access/door`.
+
+=== Payload Encoding
 
-_Coming Soon_
\ No newline at end of file
+The MQTT payload **MUST** be a UMessage that is represented as a byte array to reduce size.
```

Comment:
Thought on this one. Could we just use the UMessage Protobuf's "SerializeToString()" Method to serialize the payload before publishing it in MQTT? Would keep it consistent between different languages.

Reply:
That is a good question. I am currently using a 'write_to_bytes' protobuf function that is provided by the Protobuf crate and I believe is used by up-client-zenoh. @stevenhartley Do you have an opinion on this?

Reply:
Makes sense to use to me! But I'm not named Steven :slightly_smiling_face:

Reply:
Payload needs to be UPayload and in it there is the UPayloadFormat that says how the data is serialized. We cannot assume a given format, this is also why I suggested that we use the MQTT5 binary form so we don't have to make the MQTT PDU a string (it could be binary)

Reply:
If the Payload is UPayload, then we have to map the UAttributes to MQTT 5 User Properties.  
If the Payload is UMessage, we still have to map some UAttributes to MQTT 5 User Properties for routing purpose.

Reply:
@lynn2009 That is a valid point. The current draft has all of UMessage being passed as the mqtt payload and select UAttributes being duplicated in the MQTT 5 User Properties section. I think the current spec for a UTransport allows for a variation of what is considered payload and what is in the header as long as the UPayload is not altered. I believe this is achieved with the current proposal but I am open to enforcing all of the UAttributes into the MQTT 5 User Properties and just pass the byte serialized uPayload. It will just require some reconstruction of the UAttributes if the message needs to be sent to a different MQTT Broker (ie. getting sent off device)

Reply:
If we are able to send UAttributes in MQTT 5 User Properties, why cant we just send the UPayload in the payload? We dont need to duplicate again and increase the data over the wire.

Reply:
My original thought was to minimize the header size and not have to reconstruct the UAttributes from multiple places, but I am open to passing all the UAttributes in the header and just passing uPayload in the payload portion of the message. It will require mapping on both ends of the message (send and receive) but I dont think thats a big deal. Any concerns @matthewd0123 @PLeVasseur @lynn2009?

Reply:
@devkelley -- I was re-reading the [relevant part of the spec](https://github.com/eclipse-uprotocol/up-spec/tree/main/up-l1#21-send) to see if there was any guidance, as I recalled that there was explicit mention of putting info in transport-specific header and deduplicating in the transport payload.

I don't see any guidance on where to place the `UPayload`, but I suppose this is probably also an implementation detail of how `up-client-mqtt5-rust` would work. So long as both ends of the pipe know what to expect, should work.

As far as what's best to do -- I'm not clear on any performance gain/loss on where to put the UPayload.

tl;dr: I don't see any real problem with the proposal, but tagging @stevenhartley just to be safe :slightly_smiling_face:

Reply:
There are few reasons for the uAttributes to be in the clear as lot of commercial components may have to handle them. Even some components in cloud either might want to consume these uAttributes without having to deserialize uMessage proto. So another option is good where the uMessage has uHeaders in json and uPayload is proto. That probably is better option.

Reply:
I agree that being clear where uAttributes are in the transport message is important and I think it does make sense to keep the uAttributes together in an easily accessible location in the transport message to minimize routing work. My knowledge is lacking a bit here so I am unsure what the performance hit (from device to cloud) would be to have uAttributes serialized as json in the transport message. As far as mapping the values when switching from MQTT to another transport, my understanding is that a uTransport defines its own scheme for a uMessage, so regardless of how a MQTT uTransport handles a uMessage, the transport would need to convert back to a uMessage before passing onto the next transport.

Reply:
We will have to compare these options for data on the wire.  Payload may never go over 50 bytes but the uAttributes may come to 300 bytes however we send to make routing choices. In the other option, we have performance hit with deserializing and serializing billions of messages.

Reply:
I updated the spec to at least put all of the uATtributes in the mqtt header properties as a starting place.